workload/histogram: clean prometheus metric names

otan opened the pull request:
Replace out bad characters (as determined by
prometheus.IsValidMetricName) before creating or accessing prometheus
metric names in histogram.Get. I feel there's a few floating around with
invalid names, but we should probably accept and sanitize them instead
of banning them.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66507)
<!-- Reviewable:end -->

otan committed:
workload/histogram: clean prometheus metric names

Replace out bad characters (as determined by
prometheus.IsValidMetricName) before creating or accessing prometheus
metric names in histogram.Get. I feel there's a few floating around with
invalid names, but we should probably accept and sanitize them instead
of banning them.

Release note: None

tbg commented on pkg/workload/histogram/histogram.go:
```diff
@@ -161,6 +164,9 @@ func makePrometheusLatencyHistogramBuckets() []float64 {
 
 func (w *Registry) getPrometheusHistogram(name string) prometheus.Histogram {
+	// Metric names must be sanitized or NewHistogram will panic.
+	name = invalidPrometheusMetricRe.ReplaceAllString(name, "_")
```
This puts a regexp replace into the hot path of workloads, see for example

https://github.com/cockroachdb/cockroach/blob/31847acd14ed27a340dfc620a544c3e33cbd7c9a/pkg/workload/kv/kv.go#L404-L406

so we should get to a place where we sanitize only once, when the histogram is created. I think this is straightforward: only sanitize the `Name` field in the `if !ok` branch below, and leave it alone here.

tbg reviewed (approved):
One comment about improving the perf of the sanitizing.
Also could you add a unit test?

```go
func TestHistogramWithInvalidPrometheusName(t *testing.T) {
  r := NewRegistry(/* stuff */)
  require.NotPanics(t, func() { r.GetHandle().Get("a/b*c@d") })
  // Bonus points for checking the name of the created prom histogram
}

tbg commented:
Going to push this over the finish line, since https://github.com/cockroachdb/cockroach/issues/66529 is also affected here.
